source/controllers/src/impedance/CartesianTwistController.cpp

```diff
@@ -0,0 +1,120 @@
+    angular_damping_(std::make_shared<Parameter<double>>("angular_damping", other.get_gains()(3))),
+    dissipative_ctrl_(other.dissipative_ctrl_),
+    velocity_impedance_ctrl_(other.velocity_impedance_ctrl_) {
+}
+
+CartesianTwistController& CartesianTwistController::operator=(const CartesianTwistController& other) {
+  CartesianTwistController tmp(other);
+  swap(*this, tmp);
+  return *this;
+}
+
+void CartesianTwistController::set_gains(double linear_principle_damping,
+                                         double linear_orthogonal_damping,
+                                         double angular_stiffness,
+                                         double angular_damping) {
+  set_linear_gains(linear_principle_damping, linear_orthogonal_damping);
+  set_angular_gains(angular_stiffness, angular_damping);
+}
+
+void CartesianTwistController::set_gains(const Eigen::Vector4d& gains) {
+  set_linear_gains(gains(0), gains(1));
+  set_angular_gains(gains(2), gains(3));
+}
+
+Eigen::Vector4d CartesianTwistController::get_gains() const {
+  return Eigen::Vector4d(linear_principle_damping_->get_value(),
+                         linear_orthogonal_damping_->get_value(),
+                         angular_stiffness_->get_value(),
+                         angular_damping_->get_value());
+}
+
+void CartesianTwistController::set_linear_principle_damping(double linear_principle_damping) {
+  set_linear_gains(linear_principle_damping, linear_orthogonal_damping_->get_value());
+}
+
+void CartesianTwistController::set_linear_orthogonal_damping(double linear_orthogonal_damping) {
+  set_linear_gains(linear_principle_damping_->get_value(), linear_orthogonal_damping);
+}
+
+void CartesianTwistController::set_linear_gains(double linear_principle_damping, double linear_orthogonal_damping) {
+  linear_principle_damping_->set_value(linear_principle_damping);
+  linear_orthogonal_damping_->set_value(linear_orthogonal_damping);
+
+  Eigen::VectorXd damping(6);
+  damping << linear_principle_damping, linear_orthogonal_damping, linear_orthogonal_damping, 0, 0, 0;
+  dissipative_ctrl_.set_damping_eigenvalues(damping);
+}
+
+void CartesianTwistController::set_angular_stiffness(double angular_stiffness) {
+  set_angular_gains(angular_stiffness, angular_damping_->get_value());
+}
+
+void CartesianTwistController::set_angular_damping(double angular_damping) {
+  set_angular_gains(angular_stiffness_->get_value(), angular_damping);
+}
+
+void CartesianTwistController::set_angular_gains(double angular_stiffness, double angular_damping) {
+  angular_stiffness_->set_value(angular_stiffness);
+  angular_damping_->set_value(angular_damping);
+
+  Eigen::MatrixXd k(6, 6), d(6, 6);
+  k.diagonal() << 0, 0, 0, angular_stiffness, angular_stiffness, angular_stiffness;
+  d.diagonal() << 0, 0, 0, angular_damping, angular_damping, angular_damping;
+  velocity_impedance_ctrl_ = VelocityImpedance<CartesianState>(k, d);
```

Comment:
Why do you reset the controller and not just set the `stiffness` and `damping` matrix of the previous one?

Reply:
Fair point, easy enough to fix!